### PR TITLE
feat: generate kotlin external modules from FTL tooling

### LINKED ISSUE #970: Move codegen from Maven into the `ftl` tooling.
Use the same pattern as `go-runtime/compile.Build`, which uses [scaffolder](https://github.com/TBD54566975/scaffolder) to generate stubs in Go. We want the same thing in Kotlin. Use `internal.ScaffoldZip()` rather than scaffolder directly, as it adds common functions.

The way we get the templates into the `ftl` binary differs between development mode and production. In development we load zip files directly from disk. In a production release build we use `bit` to build the zip and go:embed to embed it. See [devel.go](https://github.com/TBD54566975/ftl/blob/aat%2Fbuildengine-watch/go-runtime/compile/devel.go#L11) and [release.go](https://github.com/TBD54566975/ftl/blob/aat%2Fbuildengine-watch/go-runtime/compile/release.go#L17) for examples.

I would just shove in the entire schema as the template context, removing any modules that the target module isn't importing.

### PR DESCRIPTION
**Pull request by worstell**
fixes #970

This implementation diverges from the kotlin-runtime module generator logic in that it omits `IngressHttp` or `Json` annotations in external modules. I don't think we need those for the generated external modules.

Will follow with a change to remove the module generator entirely and stop releasing its JAR.

### CODE REVIEW

File: buildengine/build_kotlin.go [hunk 1]
```diff
@@ -72,3 +108,180 @@ func SetPOMProperties(ctx context.Context, baseDir string) error {
...
+	_ = os.RemoveAll(filepath.Join(config.Dir, "target", "generated-sources", "ftl"))
+
+	logger.Debugf("Generating external modules")
+	return internal.ScaffoldZip(kotlinruntime.ExternalModuleTemplates(), config.Dir, externalModuleContext{
+		module: module,
+		Schema: sch,
+	}, scaffolder.Functions(funcs))
+}
+
+var scaffoldFuncs = scaffolder.FuncMap{
```
Comment (alecthomas) [inline buildengine/build_kotlin.go:156]: These functions are already defined in a couple of places. Let's factor them into a single location and include them by default in `internal.ScaffoldZip`.

Comment (alecthomas) [reply buildengine/build_kotlin.go:156]: `scaffolder.Functions()` is cumulative, so the bespoke functions can still be added in each individual location.

File: buildengine/build_kotlin_test.go [hunk 2]
```diff
@@ -0,0 +1,269 @@
...
+  val dataRefArray: List<TestRequest>,
+  val map: Map<String, Long>,
+  val nestedMap: Map<String, Map<String, Long>>,
+  val dataRef: TestRequest,
+  val externalDataRef: TestRequest,
+  val any: Any,
+  val parameterizedDataRef: ParamTestData<T>,
+  val withAlias: String,
+  val unit: Unit,
+)
```
Comment (alecthomas) [inline buildengine/build_kotlin_test.go:122]: Nice.

File: kotlin-runtime/external-module-template/target.tmpl/generated-sources/ftl/{{ range .ExternalModules }}{{ push .Name . }}{{ end }}/{{ .Name | camel }}.kt [hunk 3]
```diff
@@ -0,0 +1,34 @@
+{{- $moduleName := .Name -}}
+// Code generated by FTL. DO NOT EDIT.
+{{.Comments|comment -}}
+package ftl.{{.Name}}
+{{- $imports := (.|imports) -}}
+{{- if $imports}}
+{{range $import := $imports}}
+import {{$import}}
+{{- end}}{{end}}
```
Comment (alecthomas) [inline kotlin-runtime/external-module-template/target.tmpl/generated-sources/ftl/{{ range .ExternalModules }}{{ push .Name . }}{{ end }}/{{ .Name | camel }}.kt:9]: This is quite difficult to read. Let's introduce some whitespace here between the different semantic sections. The intent of using `{{- ` and `- }}`, is to allow the template to have meaningful whitespace while still generating clean output.

File: kotlin-runtime/ftl-runtime/src/main/kotlin/xyz/block/ftl/schemaextractor/ExtractSchemaRule.kt [hunk 4]
```diff
@@ -558,7 +559,15 @@ class SchemaExtractor(
     }
 
     private fun KtDeclaration.comments(): List<String> {
-      return this.docComment?.text?.trim()?.let { listOf(it) } ?: emptyList()
+      return this.docComment?.text?.trim()?.let { text ->
+        text.removePrefix("/*")
+          .removeSuffix("*/")
+          .lines()
+          .map { it.trim().removePrefix("*").trim() }
+          .filter { it.isNotEmpty() }
```
Comment (alecthomas) [inline kotlin-runtime/ftl-runtime/src/main/kotlin/xyz/block/ftl/schemaextractor/ExtractSchemaRule.kt:567]: I don't think we want to remove empty lines in the comments do we?

**Review by alecthomas — APPROVED** [review]
Nice one Lizzy!

File: kotlin-runtime/ftl-runtime/src/main/kotlin/xyz/block/ftl/schemaextractor/ExtractSchemaRule.kt (hunk 4 above)
Comment (worstell) [reply kotlin-runtime/ftl-runtime/src/main/kotlin/xyz/block/ftl/schemaextractor/ExtractSchemaRule.kt:567]: ah the thinking here was to filter out the lines that become empty after removing the javadoc start/end markers `/**` and `*/`- but good point, we don't want to lose intentional line spacing. I've updated to only filter out the lines that contain start/end markers with nothing else

File: kotlin-runtime/external-module-template/target.tmpl/generated-sources/ftl/{{ range .ExternalModules }}{{ push .Name . }}{{ end }}/{{ .Name | camel }}.kt (hunk 3 above)
Comment (worstell) [reply kotlin-runtime/external-module-template/target.tmpl/generated-sources/ftl/{{ range .ExternalModules }}{{ push .Name . }}{{ end }}/{{ .Name | camel }}.kt]: updated! hopefully clearer chunks now

**Pull request merged**